distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+        iv.   Compute pairwise distance matrix using PHYLIP's protdist
+              function and Z-score normalize the set of pairwise distances
+              for each gene family and species;
+        v.    Add distance matrix for all pairwise distances into global
+              distance matrix storing results for all genes
+
+    2. Cluster gene families by species,
+        vi.   Compute all species sets (sets of genes whose orthologs are
+              detectable in exactly the same subset of the considered
+              species);
+        vii.  Cluster genes to each core species set cluster using the Hamming
+              distance clustering algorithm;
+        viii. Run outlier detection algorithm on each cluster (paragraph 2
+              of section 'Detecting Outlier Genes' in original paper)
+"""
+
+import sys
+import click
+import numpy
+import operator
+import threading
+import subprocess
+import traceback
+import shlex
+from os import listdir
+from os.path import join, splitext, basename
+from itertools import imap
+
+from skbio.parse.sequences import parse_fasta
+
+
+class Command(object):
+    """
+    Enables to run subprocess commands in a different thread
+    with TIMEOUT option.
+
+    Based on jcollado's solution:
+    http://stackoverflow.com/questions/1191374/subprocess-with-timeout/4825933#4825933
+    https://gist.github.com/kirpit/1306188
+    """
+    command = None
+    process = None
+    status = None
+    output, error = '', ''
+
+    def __init__(self, command):
+        if isinstance(command, basestring):
+            command = shlex.split(command)
+        self.command = command
+
+    def run(self, timeout=None, **kwargs):
+        """ Run a command then return: (status, output, error). """
+        def target(**kwargs):
+            try:
+                self.process = subprocess.Popen(self.command, **kwargs)
+                self.output, self.error = self.process.communicate()
+                self.status = self.process.returncode
+            except:
+                self.error = traceback.format_exc()
+                self.status = -1
+        # default stdout and stderr
+        if 'stdout' not in kwargs:
+            kwargs['stdout'] = subprocess.PIPE
+        if 'stderr' not in kwargs:
+            kwargs['stderr'] = subprocess.PIPE
+        # thread
+        thread = threading.Thread(target=target, kwargs=kwargs)
+        thread.start()
+        thread.join(timeout)
+        if thread.is_alive():
+            self.process.terminate()
+            thread.join()
+        return self.status, self.output, self.error
+
+
+def hamming(str1, str2):
+    "Compute the Hamming distance between two strings"
+    assert len(str1) == len(str2)
+    ne = operator.ne
+    return sum(imap(ne, str1, str2))
+
+
+def preprocess_data(working_dir,
+                    target_proteomes_dir,
+                    verbose=False):
+    """ Map each gene to sudo name (ex. 1_1 for species 1, gene 1) for easier
```

Comment:
The maximum 10 character taxon ID is the [Phylip MSA format](http://www.bioperl.org/wiki/PHYLIP_multiple_alignment_format). The species_genes naming is limited to 9999_99999 to comply with the max 10 char format.